zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -0,0 +1,727 @@
+                            randomness,
+                        ),
+                    ],
+                    [
+                        F::from(self.id as u64),
+                        F::from(TxContextFieldTag::CallDataLength as u64),
+                        F::zero(),
+                        F::from(self.call_data_length as u64),
+                    ],
+                ],
+                self.call_data
+                    .iter()
+                    .enumerate()
+                    .map(|(idx, byte)| {
+                        [
+                            F::from(self.id as u64),
+                            F::from(TxContextFieldTag::CallData as u64),
+                            F::from(idx as u64),
+                            F::from(*byte as u64),
+                        ]
+                    })
+                    .collect(),
+            ]
+            .concat()
+        }
+    }
+
+    #[derive(Debug, Default)]
+    pub struct Call<F> {
+        pub id: usize,
+        pub is_root: bool,
+        pub is_create: bool,
+        pub opcode_source: F,
+    }
+
+    #[derive(Clone, Debug, Default)]
+    pub struct ExecStep {
+        pub call_idx: usize,
+        pub rw_indices: Vec<usize>,
+        pub execution_state: ExecutionState,
+        pub rw_counter: usize,
+        pub program_counter: u64,
+        pub stack_pointer: usize,
+        pub gas_left: u64,
+        pub gas_cost: u64,
+        pub memory_size: u64,
+        pub state_write_counter: usize,
+        pub opcode: Option<OpcodeId>,
+    }
+
+    #[derive(Debug)]
+    pub struct Bytecode {
+        pub hash: Word,
+        pub bytes: Vec<u8>,
+    }
+
+    impl Bytecode {
+        pub fn new(bytes: Vec<u8>) -> Self {
+            Self {
+                hash: Word::from_big_endian(
+                    Keccak256::digest(&bytes).as_slice(),
+                ),
+                bytes,
+            }
+        }
+
+        pub fn table_assignments<'a, F: FieldExt>(
+            &'a self,
+            randomness: F,
+        ) -> impl Iterator<Item = [F; 4]> + '_ {
+            struct BytecodeIterator<'a, F> {
+                idx: usize,
+                push_data_left: usize,
+                hash: F,
+                bytes: &'a [u8],
+            }
+
+            impl<'a, F: FieldExt> Iterator for BytecodeIterator<'a, F> {
+                type Item = [F; 4];
+
+                fn next(&mut self) -> Option<Self::Item> {
+                    if self.idx == self.bytes.len() {
+                        return None;
+                    }
+
+                    let idx = self.idx;
+                    let byte = self.bytes[self.idx];
+                    let mut is_code = true;
+
+                    if self.push_data_left > 0 {
+                        is_code = false;
+                        self.push_data_left -= 1;
+                    } else if (0x60..0x80).contains(&byte) {
+                        self.push_data_left = byte as usize - 0x60 + 1;
```

Comment:
Any reason the opcode IDs aren't used here?

Reply:
Not intended to use raw byte here, let me fix it now.